File: app/src/main/java/com/futsch1/medtimer/MedicineViewModel.java
package com.futsch1.medtimer;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.futsch1.medtimer.database.Medicine;
import com.futsch1.medtimer.database.MedicineRepository;
import com.futsch1.medtimer.database.MedicineWithReminders;
import com.futsch1.medtimer.database.Reminder;
import com.futsch1.medtimer.database.ReminderEvent;

import org.jetbrains.annotations.Nullable;

import java.util.List;

public class MedicineViewModel extends AndroidViewModel {

    public final MedicineRepository medicineRepository;
    private final LiveData<List<MedicineWithReminders>> medicines;

    public MedicineViewModel(Application application) {
        super(application);
        medicineRepository = new MedicineRepository(application);
        medicines = medicineRepository.getLiveMedicines();
    }

    public LiveData<List<MedicineWithReminders>> getMedicines() {
        return medicines;
    }

    public Medicine getMedicine(int medicineId) {
        return medicineRepository.getMedicine(medicineId);
    }

    public int insertMedicine(Medicine medicine) {
        return (int) medicineRepository.insertMedicine(medicine);
    }

    public void updateMedicine(Medicine medicine) {
        medicineRepository.updateMedicine(medicine);
    }

    public void deleteMedicine(Medicine medicine) {
        medicineRepository.deleteMedicine(medicine);
    }


    public LiveData<List<Reminder>> getLiveReminders(int medicineId) {
        return medicineRepository.getLiveReminders(medicineId);
    }

    public List<Reminder> getReminders(int medicineId) {
        return medicineRepository.getReminders(medicineId);
    }

    public Reminder getReminder(int reminderId) {
        return medicineRepository.getReminder(reminderId);
    }

    public void insertReminder(Reminder reminder) {
        medicineRepository.insertReminder(reminder);
    }

    public void updateReminder(Reminder reminder) {
        medicineRepository.updateReminder(reminder);
    }

    public void deleteReminder(Reminder reminder) {
        medicineRepository.deleteReminder(reminder);
    }

    public LiveData<List<ReminderEvent>> getLiveReminderEvents(int limit, long timeStamp, boolean withDeleted) {
        return medicineRepository.getLiveReminderEvents(limit, timeStamp, withDeleted);
    }

    public void deleteReminderEvents() {
        medicineRepository.deleteReminderEvents();
    }

    public void deleteAll() {
        medicineRepository.deleteAll();
    }

    public @Nullable ReminderEvent getReminderEvent(int reminderEventId) {
        return medicineRepository.getReminderEvent(reminderEventId);
    }

    public void updateReminderEvent(ReminderEvent reminderEvent) {
        medicineRepository.updateReminderEvent(reminderEvent);
    }
}

File: app/src/main/java/com/futsch1/medtimer/MedicineViewModel.kt
package com.futsch1.medtimer

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.futsch1.medtimer.database.FullMedicineEntity
import com.futsch1.medtimer.database.MedicineRepository
import com.futsch1.medtimer.database.MedicineToTagEntity
import com.futsch1.medtimer.database.TagEntity
import com.futsch1.medtimer.medicine.tags.TagFilterStore
import com.futsch1.medtimer.model.ScheduledReminder
import com.futsch1.medtimer.model.reminderevent.ReminderEvent
import com.futsch1.medtimer.preferences.PersistentDataDataSource
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.shareIn
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.util.stream.Collectors
import javax.inject.Inject

@HiltViewModel
class MedicineViewModel @Inject constructor(
    persistentDataDataSource: PersistentDataDataSource,
    private val medicineRepository: MedicineRepository,
) : ViewModel() {
    private val liveMedicines = medicineRepository.medicinesFlow

    val validTagIds = MutableStateFlow<Set<Int>?>(null)
    val tagFilterStore = TagFilterStore(persistentDataDataSource, validTagIds)
    private var medicineToTags: List<MedicineToTagEntity> = emptyList()
    private val liveTags: StateFlow<List<TagEntity>> = medicineRepository.tagsFlow.stateIn(
        viewModelScope, SharingStarted.Eagerly, emptyList()
    )

    private val _scheduledReminders = MutableStateFlow<List<ScheduledReminder>>(emptyList())

    val medicines: StateFlow<List<FullMedicineEntity>> = combine(liveMedicines, validTagIds) { medicines, tagIds ->
        getFiltered(medicines, tagIds ?: emptySet())
    }.stateIn(viewModelScope, SharingStarted.Eagerly, emptyList())

    val scheduledReminders: SharedFlow<List<ScheduledReminder>> = combine(_scheduledReminders, validTagIds) { reminders, tagIds ->
        getFiltered(reminders, tagIds ?: emptySet())
    }.shareIn(viewModelScope, SharingStarted.Eagerly, replay = 1)

    init {
        viewModelScope.launch {
            medicineRepository.getMedicineToTagsFlow().collect {
                medicineToTags = it
            }
        }

        viewModelScope.launch {
            medicineRepository.tagsFlow.collect {
                tagFilterStore.filterForDeletedTags(it)
            }
        }
    }

    private fun <T : Any> getFiltered(
        liveData: List<T>,
        validTagIds: Set<Int>
    ): List<T> {
        if (validTagIds.isEmpty()) {
            return liveData
        }
        return liveData.stream()
            .filter { medicine ->
                filterMedicineId(
                    getId(medicine),
                    validTagIds
                )
            }
            .collect(Collectors.toList())
    }

    fun filterEvents(events: List<ReminderEvent>): List<ReminderEvent> {
        return getFilteredEvents(events, validTagIds.value ?: setOf(), liveTags.value)
    }

    fun filterMedicines(medicines: List<FullMedicineEntity>): List<FullMedicineEntity> {
        return getFiltered(medicines, validTagIds.value ?: emptySet())
    }

    fun tagFilterActive(): Boolean {
        return validTagIds.value?.isNotEmpty() ?: return false
    }

    private fun getFilteredEvents(
        liveData: List<ReminderEvent>,
        validTagIds: Set<Int>,
        liveTags: List<TagEntity>
    ): List<ReminderEvent> {
        if (validTagIds.isEmpty()) {
            return liveData
        }
        // Get all valid tag names from the list of valid IDs
        val validTagNames =
            liveTags.stream().filter { tag -> validTagIds.contains(tag.tagId) }
                .map { tag -> tag.name }.collect(Collectors.toSet())
        // Now filter all reminder events and check if they contain any of the valid tags
        return liveData.stream().filter { reminderEvent ->
            validTagNames.stream().filter { reminderEvent.tags.contains(it) }.count() > 0
        }.collect(Collectors.toList())
    }

    private fun getId(medicineWithReminder: Any): Int {
        return if (medicineWithReminder is FullMedicineEntity) {
            medicineWithReminder.medicine.medicineId
        } else {
            (medicineWithReminder as ScheduledReminder).medicine.medicine.medicineId
        }
    }

    private fun filterMedicineId(medicineId: Int, validTagIds: Set<Int>? = null): Boolean {
        if (validTagIds.isNullOrEmpty() || medicineToTags.isEmpty()) {
            return true
        }
        val medicineTags: List<Int> = medicineToTags.stream()
            .filter { medicineToTag -> medicineToTag.medicineId == medicineId }
            .map { medicineToTag -> medicineToTag.tagId }
            .collect(Collectors.toList())
        for (tagId in validTagIds) {
            if (medicineTags.contains(tagId)) {
                return true
            }
        }
        return false
    }

    fun getLiveReminderEvents(
        timeStamp: Long,
        statusValues: List<ReminderEvent.ReminderStatus> = ReminderEvent.allStatusValues
    ): Flow<List<ReminderEvent>> {
        return combine(
            medicineRepository.getReminderEventsFlow(timeStamp, statusValues),
            validTagIds,
            liveTags
        ) { events, tagIds, tags ->
            if (tagIds.isNullOrEmpty()) events else getFilteredEvents(events, tagIds, tags)
        }.shareIn(viewModelScope, SharingStarted.Eagerly, replay = 1)
    }

    fun setScheduledReminders(scheduledReminders: List<ScheduledReminder>) {
        _scheduledReminders.value = scheduledReminders
    }
}
